src/strawman/collections/CollectionStrawMan6.scala

```diff
@@ -0,0 +1,980 @@
+ *
+ *         IterableOps, IterableMonoTransforms, IterablePolyTransforms, IterableLike
+ *         SeqMonoTransforms, SeqLike
+ *
+ *  4. Concrete collection types:
+ *
+ *         List, LazyList, ListBuffer, ArrayBuffer, ArrayBufferView, StringView, ArrayView
+ *
+ *  5. Decorators for existing types
+ *
+ *         StringOps, ArrayOps
+ *
+ *  6. Related non collection types:
+ *
+ *         Iterator, StringBuilder
+ *
+ *  Operations covered in this strawman:
+ *
+ *   1. Abstract operations, or expected to be overridden:
+ *
+ *      For iterables:
+ *
+ *         iterator, fromIterable, fromIterableWithSameElemType, knownLength, className
+ *
+ *      For sequences:
+ *
+ *         apply, length
+ *
+ *      For buildables:
+ *
+ *         newBuilder
+ *
+ *      For builders:
+ *
+ *         +=, result
+ *
+ *   2. Utility methods, might be overridden for performance:
+ *
+ *      Operations returning not necessarily a collection:
+ *
+ *         foreach, foldLeft, foldRight, indexWhere, isEmpty, head, size, mkString
+ *
+ *      Operations returning a collection of a fixed type constructor:
+ *
+ *         view, to, toArray, copyToArray
+ *
+ *      Type-preserving generic transforms:
+ *
+ *         filter, partition, take, drop, tail, reverse
+ *
+ *      Generic transforms returning collections of different element types:
+ *
+ *         map, flatMap, ++, zip
+ */
+object CollectionStrawMan6 extends LowPriority {
+
+  /* ------------ Base Traits -------------------------------- */
+
+  /** Iterator can be used only once */
+  trait IterableOnce[+A] {
+    def iterator: Iterator[A]
+  }
+
+  /** Base trait for instances that can construct a collection from an iterable */
+  trait FromIterable[+C[X] <: Iterable[X]] {
+    def fromIterable[B](it: Iterable[B]): C[B]
+  }
+
+  /** Base trait for companion objects of collections */
+  trait IterableFactory[+C[X] <: Iterable[X]] extends FromIterable[C] {
+    def empty[X]: C[X] = fromIterable(View.Empty)
+    def apply[A](xs: A*): C[A] = fromIterable(View.Elems(xs: _*))
+  }
+
+  /** Base trait for generic collections */
+  trait Iterable[+A] extends IterableOnce[A] with IterableLike[A, Iterable] {
+    /** The collection itself */
+    protected def coll: this.type = this
+ }
+
+  /** Base trait for sequence collections */
+  trait Seq[+A] extends Iterable[A] with SeqLike[A, Seq] {
+    def apply(i: Int): A
+    def length: Int
+  }
+
+  /** Base trait for linearly accessed sequences */
+  trait LinearSeq[+A] extends Seq[A] with SeqLike[A, LinearSeq] { self =>
+
+    /** To be overridden in implementations: */
+    def isEmpty: Boolean
+    def head: A
+    def tail: LinearSeq[A]
+
+    def iterator = new Iterator[A] {
+      private[this] var current: Seq[A] = self
+      def hasNext = !current.isEmpty
+      def next = { val r = current.head; current = current.tail; r }
+    }
+
+    def length: Int = iterator.length
+
+    @tailrec final def apply(n: Int): A =
```

Comment:
I'm also wondering if there should be a `LinearSeqLike` for these overrides, instead of putting them into `LinearSeq`. Keeping all operations in the "Like" traits seems more consistent.